src/engraving/dom/dynamic.cpp

```diff
@@ -942,4 +969,120 @@ String Dynamic::screenReaderInfo() const
+
+void Dynamic::drawEditMode(muse::draw::Painter* p, EditData& ed, double currentViewScaling)
+{
+    if (ed.editTextualProperties) {
+        TextBase::drawEditMode(p, ed, currentViewScaling);
+    } else {
+        EngravingItem::drawEditMode(p, ed, currentViewScaling);
+    }
+}
+
+//---------------------------------------------------------
+//   hasLeftHairpin
+//---------------------------------------------------------
+
+bool Dynamic::hasLeftGrip() const
+{
+    if (segment()->tick().isZero()) {
+        return false; // Don't show the left grip for the leftmost dynamic with tick zero
+    }
+    return m_leftHairpin == nullptr;
+}
+
+//---------------------------------------------------------
+//   hasRightHairpin
+//---------------------------------------------------------
+
+bool Dynamic::hasRightGrip() const
+{
+    return m_rightHairpin == nullptr;
+}
+
+//---------------------------------------------------------
+//   findAdjacentHairpins
+//---------------------------------------------------------
+
+void Dynamic::findAdjacentHairpins()
+{
+    m_leftHairpin = nullptr;
+    m_rightHairpin = nullptr;
+
+    const Fraction tick = segment()->tick();
+    const int intTick = tick.ticks();
+
+    const auto& spanners = score()->spannerMap().findOverlapping(intTick - 1, intTick + 1);
+    for (auto i : spanners) {
+        Spanner* sp = i.value;
+        if (sp->track() == track() && sp->isHairpin()) {
+            Hairpin* hp = toHairpin(sp);
+            if (hp->tick() == tick) {
+                m_rightHairpin = hp;
+            } else if (hp->tick2() == tick) {
+                m_leftHairpin = hp;
+            }
+        }
+    }
+
+    return;
+}
+
+//---------------------------------------------------------
+//   gripsCount
+//---------------------------------------------------------
+
+int Dynamic::gripsCount() const
+{
+    if (empty()) {
+        return 0;
+    }
+
+    if (hasLeftGrip() && hasRightGrip()) {
+        return 2;
+    } else if (hasLeftGrip() ^ hasRightGrip()) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+//---------------------------------------------------------
+//   gripsPositions
+//---------------------------------------------------------
+
+std::vector<PointF> Dynamic::gripsPositions(const EditData&) const
+{
+    const LayoutData* ldata = this->ldata();
+    const PointF pp(pagePos());
+    double md = score()->style().styleS(Sid::hairpinMinDistance).val() * spatium(); // Minimum distance between dynamic and grip
+
+    PointF leftOffset(-ldata->bbox().width() / 2 - md + m_leftDragOffset, -11.408);
+    PointF rightOffset(ldata->bbox().width() / 2 + md + m_rightDragOffset, -11.408);
+
+    if (hasLeftGrip() && !hasRightGrip()) {
+        return { pp + leftOffset };
+    }
+    if (!hasLeftGrip() && hasRightGrip()) {
+        return { pp + rightOffset };
+    }
+    return { pp + leftOffset, pp + rightOffset };
+}
+
+RectF Dynamic::adjustedBoundingRect() const
```

Comment:
I have two concerns about this method:
1. We should think together about whether we really need this, because it's introducing a special case for dynamics and special cases are not always desirable. Where possible, I prefer generalisation. @mike-spa Do you have ideas about this?
2. In some sense, this method hardcodes the height, albeit spatium-dependent. This may not work well for increased/decreased font sizes.

Reply:
@cbjeukendrup One of the main reasons I added this method was to ensure a fixed y-positioning for the popup. This method is actually used in two places: one is `TextBase::drawEditMode`, which draws the rectangle around the dynamic element, and the other place is `updateContainerPosition` in `ElementPopupLoader.qml` where it is used as `elementRect`, which seems to be responsible for setting the position of the popup. Previously, it used `canvasBoundingRect` as `elementRect`.

So, if I remove this method and use `canvasBoundingRect` as earlier it was causing these two issues:
1. When we open an empty popup using the shortcut and hover over any dynamic, the popup position doesn't change because it uses the `canvasBoundingRect` of an empty dynamic, as shown below.

https://github.com/user-attachments/assets/64ff3659-3f01-481c-b7c8-90513cb19fc6

2. I was able to figure out the fix for the above (by emitting `itemRectChanged` in `showPreview` and `hidePreview` methods) but it caused another problem. If we hover over a dynamic while slightly zoomed in, the popup's position will change due to a change in `canvasBoundingRect`, causing the state to hover out which results in the popup moving up and down, making it impossible to select the dynamic. You can see in the video below what I mean.

https://github.com/user-attachments/assets/76f819df-5752-4901-862d-b0cd9c53ffe4

Fixing the height of `elementRect` by using `adjustBoudingRect` was one way I was able to resolve this behaviour.